Mapgen palettes

Coolthulhu opened the pull request:
Implements #19578

Works as described in the issue.
Palettes can be used with inline declarations, in this case palettes are applied first, then inline declarations overwrite them.

Changed the game basement to show it in action.

tswood commented on src/mapgen.cpp:
```diff
@@ -1294,6 +1294,132 @@ void mapgen_function_json::load_place_mapings( JsonObject &jo, const std::string
+    }
+}
+
+mapgen_palette mapgen_palette::load_internal( JsonObject &jo, const std::string &, bool require_id, bool allow_recur )
```
Am I crazy or is the second parameter missing a name? The header shows `src`, matching the name of all instances passed in currently.

Coolthulhu replied:
Yes.
Name only needs to be supplied if the parameter is used. Otherwise it will produce a compiler warning about unused parameters.
Our mapgen doesn't support passing source yet.

Reclusive-reptile commented:
The sink and toilet are still embedded in solid rock due to the fill terrain not being overridden in the terrain definition.

codemime commented on src/init.cpp:
```diff
@@ -239,6 +239,7 @@ void DynamicDataLoader::initialize()
     add( "overlay_order", &load_overlay_ordering );
     add( "mission_definition", []( JsonObject &jo, const std::string &src ) { mission_type::load_mission_type( jo, src ); } );
     add( "harvest", []( JsonObject &jo, const std::string &src ) { harvest_list::load( jo, src ); } );
+    add( "palette", []( JsonObject &jo, const std::string &src ) { mapgen_palette::load( jo, src ); } );
```
Why not just `add( "palette", &mapgen_palette::load );`? I assume that would work.

codemime commented on src/mapgen.cpp:
```diff
@@ -1294,6 +1294,132 @@ void mapgen_function_json::load_place_mapings( JsonObject &jo, const std::string
+        if( allow_recur ) {
+            auto pals = jo.get_string_array( "palettes" );
+            for( auto &p : pals ) {
+                new_pal.add( p );
```
Can't this line fail under certain (valid) circumstances? It expects a loaded palette, but the load order isn't strictly guaranteed. IMHO It would be more robust in the finalization section where everything's already loaded.

Coolthulhu replied:
I'm pretty sure that load order enforces regular static data before mapgen, as mapgen is actually processed on demand.
Meaning that by the time inline palettes are processed, all named palettes are loaded.

I intend to follow up on it to allow nested named palettes, but at the moment trying to nest a named palette will simply fail (because `allow_recur == false`).

codemime replied:
> I'm pretty sure that load order enforces regular static data before mapgen, as mapgen is actually processed on demand.

Yes, you're right. Apologies, I've missed that. It gets loaded from `finalize_loaded_data()`, although not on demand.

Coolthulhu commented:
Fixed the solid rock sink and toilet.
Removed `fill_ter` from basement definitions - it's safer to have it warn on missing terrain instead of just not filling it.

mugling commented:
I'm still wary similar to @codemime's comment from #19578 that this could lead to obfuscation if we aren't careful

Coolthulhu commented:
It doesn't need to be the standard. "One shot" mapgen works fine with inline, named palettes shine when you actually need them to avoid repeating same exact named palette over and over, as it is the case in multi-tile mapgen objects. And multi-tile mapgen objects are already multi-file anyway, since you need the overmap file to piece them together.

It could be more localized by specifying palettes in the same files they are used in.

Coolthulhu commented:
Solved conflicts.

Requesting comments - is there something more that needs to be added here?
This PR is supposed to be mostly a way to get #19572 manageable. Do we still care? Is @acidia still with us?

BorkBorkGoesTheCode commented:
>Fixed the solid rock sink and toilet.
 Removed  fill_ter  from basement definitions - it's safer to have it warn on missing terrain instead of just not filling it.

Until the debugmessages are fixed, more warnings may harm gameplay.

John-Candlebury commented:
>Do we still care?

I do, since I dont think I can start working on #20050 without this being implemented.

acidia commented:
I usually have a few less hectic months a years to contribute so it might be a bit before I can make full use of this but it is fairly critical to keep the json size from exploding on any large additions.  I should have a bit of time in two weeks to update #19572 to something more reasonable.
